### PR TITLE
adds basic cypress test

### PR DESCRIPTION
**Pull request by thoniTUB**

### CODE REVIEW

File: frontend/cypress/integration/user_frontend_test.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
+/// <reference types="cypress" />
+
+context('Visit conquery as users with different permissions', () => {
```
Comment (Kadrian) [inline frontend/cypress/integration/user_frontend_test.js:3]: We're using prettier to auto-format js / ts code on save. Could you configure your editor to use prettier? (and use our installed / configured prettier – see `package.json`)

File: frontend/cypress.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (Kadrian) [inline frontend/cypress.json:1]: Is this file required?

File: frontend/cypress/integration/user_frontend_test.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            ).as('meInfo')
+    
+            cy.visit("http://localhost:3000/?access_token=user.user2")
+    
+            cy.wait(['@config', '@datasets', '@meInfo'])
+        })
+
+        
+        it('Can see dataset', () => {
+            cy.get('#downshift-0-input').invoke('attr','value').should('eq',"Dataset1")
```
Comment (Kadrian) [inline frontend/cypress/integration/user_frontend_test.js:75]: Time for me to start adding proper test ids in the future

File: scripts/run_backend_cypress.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+config=$(cat << EOF
+ {
+    "server": {
+        "applicationConnectors": [
+            {
+                "type": "http",
+                "port": "8080"
+            }
+        ],
+		"adminConnectors": [
```
Comment (Kadrian) [inline scripts/run_backend_cypress.sh:13]: How about cleaning the indentation a bit here? :-D

**Review by Kadrian — COMMENTED** [review]
Cool draft so far! Would it make sense to add some instruction to README.md as well?

File: frontend/cypress/integration/user_frontend_test.js (hunk 1 above)
Comment (thoniTUB) [reply frontend/cypress/integration/user_frontend_test.js]: I added that to the ReadMe aswell

File: frontend/cypress.json (hunk 2 above)
Comment (thoniTUB) [reply frontend/cypress.json:1]: Not jet, but thats where the `baseUrl` is supposed to be i guess ;)

File: frontend/README.md [hunk 5]
```diff
@@ -80,6 +80,15 @@ When queried for login:
...
 
+**Code Formating**  [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
+
+The frontend JavaScript/TypeScript code is formatted using `prettier`.
+It can be invoked on the command line through `yarn`:
+```
+yarn prettier --write /path/to/file
+```
+or convieniently through a plugin in VS Code (see https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode).
+
```
Comment (Kadrian) [inline frontend/README.md:91]:
````suggestion
**Code formatting**  [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

The frontend TypeScript code is formatted using `prettier`.

We recommend you configure your editor to auto-format on save. If you're using VS-Code, for example, there's a plugin: [Prettier – Code formatter](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode).

You could also invoke prettier on the command line:
```
yarn prettier --write /path/to/file
```

````

File: frontend/src/js/pane/Pane.tsx [hunk 6]
```diff
@@ -24,16 +24,23 @@ interface PropsT {
...
     <Root className={className}>
       <Container>
-        <PaneTabNavigation tabs={tabs} paneType={paneType} />
-        <Container>{children}</Container>
+        <PaneTabNavigation
+          tabs={tabs}
+          paneType={paneType}
+          dataTestId={dataTestId}
+        />
+        <Container data-test-id={dataTestId + "Container"}>
```
Comment (Kadrian) [inline frontend/src/js/pane/Pane.tsx:41]:
```suggestion
        <Container data-test-id={dataTestId + "-container"}>
```

File: frontend/src/js/query-runner/QueryRunnerButton.tsx [hunk 7]
```diff
@@ -66,7 +66,7 @@ const QueryRunnerButton = ({
   const icon = getIcon(isStartStopLoading, isQueryRunning);
 
   return (
-    <StyledBasicButton type="button" onClick={onClick} disabled={disabled}>
+    <StyledBasicButton type="button" onClick={onClick} disabled={disabled} data-test-id="queryRunnerButton">
```
Comment (Kadrian) [inline frontend/src/js/query-runner/QueryRunnerButton.tsx:69]:
```suggestion
    <StyledBasicButton type="button" onClick={onClick} disabled={disabled} data-test-id="query-runner-button">
```

File: frontend/cypress/integration/user_frontend_test.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        url: "/api/me",
+      }).as("meInfo");
+
+      cy.visit("http://localhost:3000/?access_token=user.user1");
+
+      cy.wait(["@config", "@datasets", "@meInfo"]);
+    });
+
+    it("Cannot see concepts", () => {
+      cy.get('[data-test-id="leftPane"]').contains("Konzepte");
```
Comment (Kadrian) [inline frontend/cypress/integration/user_frontend_test.js:25]:
```suggestion
      cy.get('[data-test-id="left-pane"]').contains("Konzepte");
```

File: frontend/cypress/integration/user_frontend_test.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+      cy.contains("Keine Konzepte");
+    });
+
+    it("Cannot see queries", () => {
+      cy.get('[data-test-id="leftPane"]').contains("Anfragen").click();
```
Comment (Kadrian) [inline frontend/cypress/integration/user_frontend_test.js:31]:
```suggestion
      cy.get('[data-test-id="left-pane"]').contains("Anfragen").click();
```

File: frontend/cypress/integration/user_frontend_test.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+      cy.contains("Keine Anfragen gefunden");
+    });
+
+    it("Cannot see forms", () => {
+      cy.get('[data-test-id="leftPane"]').contains("Formulare").click();
```
Comment (Kadrian) [inline frontend/cypress/integration/user_frontend_test.js:37]:
```suggestion
      cy.get('[data-test-id="left-pane"]').contains("Formulare").click();
```

File: frontend/cypress/integration/user_frontend_test.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        url: "/api/me",
+      }).as("meInfo");
+
+      cy.visit("http://localhost:3000/?access_token=user.user2");
+
+      cy.wait(["@config", "@datasets", "@meInfo"]);
+    });
+
+    it("Can see dataset", () => {
+      cy.get('[data-test-id="datasetSelector"]')
```
Comment (Kadrian) [inline frontend/cypress/integration/user_frontend_test.js:64]:
```suggestion
      cy.get('[data-test-id="dataset-selector"]')
```

File: frontend/cypress/integration/user_frontend_test.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        .should("eq", "Dataset1");
+    });
+
+    it("Can see concepts", () => {
+      cy.contains("Concept1");
+    });
+
+    it("Can execute query", () => {
+
+      cy.get('[data-test-id="rightPaneContainer"]')
```
Comment (Kadrian) [inline frontend/cypress/integration/user_frontend_test.js:76]:
```suggestion
      cy.get('[data-test-id="right-pane-container"]')
```

File: frontend/cypress/integration/user_frontend_test.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+      cy.contains("Concept1").trigger("dragstart").trigger("dragleave");
+      cy.get('@queryEditor')
+        .trigger("dragenter")
+        .trigger("dragover")
+        .trigger("drop")
+        .trigger("dragend");
+
+      cy.get('@queryEditor')
+        .find('[data-test-id="queryRunnerButton"]').click()
```
Comment (Kadrian) [inline frontend/cypress/integration/user_frontend_test.js:89]:
```suggestion
        .find('[data-test-id="query-runner-button"]').click()
```

**Comment by Kadrian** [comment]
Here's a PR that merges in develop and resolves the conflicts in package.json / yarn.lock: https://github.com/bakdata/conquery/pull/2367

File: frontend/package.json [hunk 14]
```diff
@@ -115,10 +116,6 @@
     "ts-node": "^10.5.0",
     "typescript": "^4.5.5"
   },
-  "resolutions": {
```
Comment (Kadrian) [inline frontend/package.json:118]: Did you remove this intentionally?

Comment (thoniTUB) [reply frontend/package.json:118]: Yes, I wanted to ask you about that, since it caused a problem in the install stage, when i invoked yarn with `--frozen-lockfile`:
```
error Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.
```
Does these packages need to have fixed versions?

Comment (Kadrian) [reply frontend/package.json:118]: That's an unrelated error. Yes those versions should be fixed. I'd suggest you
- add them back
- run `yarn` without `--frozen-lockfile`
- commit the changes to yarn.lock

Comment (thoniTUB) [reply frontend/package.json:118]: The thing is, that yarn.lock won't change if i run it without `--frozen-lockfile`. I tried it, and run `yarn --ignore-platform --frozen-lockfile` afterwards again and the Problem remains. Or do you suggest, that `--frozen-lockfile` should never be used?

Comment (thoniTUB) [reply frontend/package.json:118]: I got the idea from this: https://github.com/yarnpkg/yarn/issues/4778#issuecomment-339448000

Comment (Kadrian) [reply frontend/package.json:118]: We're only using `--frozen-lockfile` in CI, so we can make sure that the dependencies installed in an automated CI build are exactly the ones we expect

Comment (thoniTUB) [reply frontend/package.json:118]: That was my intention. You can try to run `yarn --ignore-platform --frozen-lockfile` on the `develop` branch. That's where i am getting the error too. For some reason the `Test Frontend`-Workflow does not use the frozen-flag. Why is that so?

Comment (thoniTUB) [reply frontend/package.json:118]: Out of curiosity: what makes these Package version so important to stay at that fixed version?

Comment (thoniTUB) [reply frontend/package.json:118]: @Kadrian I kept the resolutions and adjusted the version of babel-loader in the dependencies to be the same as in the resolutions. This works for me now.

File: backend/src/main/java/com/bakdata/conquery/models/config/ColumnConfig.java [hunk 15]
```diff
@@ -77,40 +77,34 @@
 	 *
 	 * Also Name of output column for {@link FrontendConfig.UploadConfig#getIdResultInfos()}, ergo output csv-Columns.
 	 */
-	@InternalOnly
 	private String field;
```
Comment (thoniTUB) [inline backend/src/main/java/com/bakdata/conquery/models/config/ColumnConfig.java:80]: @awildturtok Weißt du warum hier `@InternalOnly` dranne war? Das Problem war hier das die Werte beim kopieren der Config in Standalone nicht mitgenommen wurden

File: frontend/cypress/support/backend_config.json [hunk 16]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (thoniTUB) [inline frontend/cypress/support/backend_config.json:1]: this is empty for now, but in case that a configuration is needed the place is here

**Review by Kadrian — CHANGES_REQUESTED** [review]
Looking pretty good! But I tried to run it locally and it doesn't work yet – I left you some comments

File: scripts/run_frontend_cypress.sh [hunk 17]
```diff
@@ -0,0 +1,8 @@
+#!/bin/env bash
+proj_dir=$(readlink -f `dirname $0`/..)
+
+pushd "$proj_dir/frontend"
+
+npx serve -s build -l 3000
```
Comment (Kadrian) [inline scripts/run_frontend_cypress.sh:6]: What does this do, by the way? Can / should we build & start the frontend using the Dockerfile, in a container?

File: README.md [hunk 18]
```diff
@@ -18,10 +18,36 @@ Conquery supplies a powerful interface to group event types in a hierarchical *c
...
+- Maven 3 (optional for building)
+- Java JDK 11
+- Node.js 16 + Yarn
+
+#### **Steps**
+To test frontend and backend together you can start the setup that is used for end-to-end tests.
+
+First build the backend using `scripts/build_version.sh` or download a JAR from the [release page](https://github.com/bakdata/conquery/releases).
+
+Then start frontend and backend by running `scripts/run_backend_cypress.sh`.
```
[the remark this answers is not in the document]
Comment (thoniTUB) [reply README.md:38]: I'll revise that, thank you!
There is a script call `run_conquery_cypress.sh` that runs both

File: scripts/run_frontend_cypress.sh (hunk 17 above)
Comment (thoniTUB) [reply scripts/run_frontend_cypress.sh:6]: This spins up an adhoc HTTP file-server serving files from the folder `build` (similar to pythons `python3 -m http.server` ). I will add that to the documentation, that the frontend need to be build beforehand. `npx` download `serve` incase it was not installed. I wouldn't add `serve` as a dev dependency, since no actual dev dependencies are needed here. 

Using Docker poses an extra requirement towards the need to having Docker installed and thus having to invoke a `sudo`-action at least once. All other tools like maven, java, node usually don't need `sudo`.

So if you just want to test the app or run the cypress test, you couldn't do so in a restricted environment.

File: scripts/load_test_data_cypress.sh [hunk 19]
```diff
@@ -0,0 +1,21 @@
+#!/bin/env bash
+proj_dir=$(readlink -f `dirname $0`/..)
```
[the remark this answers is not in the document]
Comment (thoniTUB) [reply scripts/load_test_data_cypress.sh:2]: I guess `-f` can be removed. It just follows symlinks

File: scripts/run_frontend_cypress.sh [hunk 20]
```diff
@@ -0,0 +1,8 @@
...
+
+popd
```
[the remark this answers is not in the document]
Comment (thoniTUB) [reply scripts/run_frontend_cypress.sh:8]: In my opinion, the scripts behavior is more consistent this way. I would not expect that my current directory changes after i run a random script/command

File: scripts/run_frontend_cypress.sh (hunk 17 above)
Comment (Kadrian) [reply scripts/run_frontend_cypress.sh:6]: > Using Docker poses an extra requirement towards the need to having Docker installed and thus having to invoke a sudo-action at least once. All other tools like maven, java, node usually don't need sudo.

Alright, I mean we should try to replicate a real world as close as possible but I'd also prefer a simple setup.

But then I'd vote to not add another dependency but just use `server.js`, which was written precicely to serve the built frontend from `./build`

File: frontend/src/js/app/RightPane.tsx [hunk 21]
```diff
@@ -12,10 +12,11 @@ interface PropsT {
   tabs: TabT[];
 }
 
+// Shrink to size 0 so jQuery/cypress can differentiate between :hidden and :visible
 const Tab = styled("div")<{ isActive: boolean }>`
-  height: 100%;
+  height: ${(isActive) => (isActive ? "100%" : 0 )};
```
[the remark this answers is not in the document]
Comment (thoniTUB) [reply frontend/src/js/app/RightPane.tsx:17]: Cypress uses jQuery under the hood, but i could find a way to filter DOM elements by their style/css directly. The most convienient way seem to use jQueries `:visible/:hidden` selector, which only looks at widths and heights of the elements:
https://api.jquery.com/visible-selector/

[the remark this answers is not in the document]
Comment (thoniTUB) [reply frontend/src/js/app/RightPane.tsx:17]: Do you mean the container that holds the Dropzone and the QueryEditor? We can do that, but from a user handling perspective the visibility "flag" or some thing like it is better. 

I haven't tested it, but i imagine that if i reference the pane by it's id rather than its visibility, it is possible that can d'n'd elements in the QueryEditor eventhough its tab isn't actually in the foreground. This way an error might slip through.

What do you think of that?

[the remark this answers is not in the document]
Comment (thoniTUB) [reply frontend/src/js/app/RightPane.tsx:17]: I think i tried that without success, but basically that is what is done in this video: https://www.youtube.com/watch?v=LxkrhUEE2Qk So I'll try that again, i must have missed something the first time

[the remark this answers is not in the document]
Comment (thoniTUB) [reply frontend/src/js/app/RightPane.tsx]: I was able to fix that :) 
The selector is got simpler now

File: scripts/load_test_data_cypress.sh [hunk 22]
```diff
@@ -0,0 +1,21 @@
+#!/bin/env bash
```
[the remark this answers is not in the document]
Comment (thoniTUB) [reply scripts/load_test_data_cypress.sh:1]: @Kadrian würde bei dir auch `#!/bin/bash` gehen?

File: scripts/load_test_data_cypress.sh (hunk 19 above)
[the remark this answers is not in the document]
Comment (thoniTUB) [reply scripts/load_test_data_cypress.sh:2]: Okay, `-f` is needed, since is outputs the actual path. 

Unfortunately I haven't found a good portable way. There is however a greadlink in homebrew coreutils. Is that an option?

File: scripts/load_test_data_cypress.sh (hunk 22 above)
[the remark this answers is not in the document]
Comment (Kadrian) [reply scripts/load_test_data_cypress.sh:1]: Yep!

File: scripts/load_test_data_cypress.sh (hunk 19 above)
[the remark this answers is not in the document]
Comment (Kadrian) [reply scripts/load_test_data_cypress.sh:2]: I've added a suggestion that works for me in here https://github.com/bakdata/conquery/pull/2400 (simple avoids using readlink)

File: .github/workflows/test-cypress.yml [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+          TSC_COMPILE_ON_ERROR: true
+          REACT_APP_API_URL: "http://localhost:8080"
+          REACT_APP_DISABLE_LOGIN: true
+          REACT_APP_LANG: de
+          REACT_APP_IDP_ENABLE: false
+          REACT_APP_IDP_REALM: Myrealm
+          REACT_APP_IDP_CLIENT_ID: frontend
+        with:
+          working-directory: ./frontend
+          build: yarn run react-app-rewired build
```
Comment (thoniTUB) [inline .github/workflows/test-cypress.yml:38]: Anstatt des commands hier ein script benutzen, damit das bauen lokal gleich funktioniert

**Comment by Kadrian** [comment]
Todo: 
- remove readlink

**Comment by thoniTUB** [comment]
add separate .env.e2e that is sourced for the cypress frontend

**Comment by Kadrian** [comment]
Another todo:
- [ ] avoid adjusting different kinds of pom.xml versions when building for cypress

<details>
<summary>Screenshots showing the issue</summary>
Bildschirmfoto 2022-03-22 um 14 21 58
Bildschirmfoto 2022-03-22 um 14 21 49
</details>

**Comment by Kadrian** [comment]
are there some chmod +x missing?

Bildschirmfoto 2022-03-28 um 10 51 27

File: README.md [hunk 24]
```diff
@@ -16,12 +14,54 @@ Event data sets typically associate events with a certain subject (i.e. a person
...
 This repository includes the Conquery frontend along with a non-functional backend. It provides a set of example concept trees to demonstrate the capabilities of the UI: The example's use case is to search for groups of actors who appeared in movies of the same genre or received the same award.
 
 Check the README in `/frontend` for details.
 
+### Frontend + Backend
+
+#### Steps
+To test frontend and backend together you can start the setup that is used for end-to-end tests.
+
+First build the backend using `conquery/scripts/build_version.sh` or download a JAR from the [release page](https://github.com/bakdata/conquery/releases) and place it in `conquery/executable/target/`.
```
Comment (Kadrian) [inline README.md:36]:
```suggestion
First build the backend using `./scripts/build_version.sh` or download a JAR from the [release page](https://github.com/bakdata/conquery/releases) and place it in `./executable/target/`.
```

File: README.md [hunk 25]
```diff
@@ -16,12 +14,54 @@ Event data sets typically associate events with a certain subject (i.e. a person
...
+
+Build the frontend by running:
+```
+$ cd frontend
+$ cp .env.example .env
+$ yarn build
+```
+
+You can then run `conquery/scripts/run_conquery_cypress.sh` to start frontend and backend, and also load the test data required by cypress end-to-end test or you can run `conquery/scripts/run_backend_cypress.sh` and `conquery/scripts/run_frontend_cypress.sh` separately without loading any data.
```
Comment (Kadrian) [inline README.md:45]:
```suggestion
You can then run `./scripts/run_conquery_cypress.sh` to start frontend and backend, and also load the test data required by cypress end-to-end test or you can run `./scripts/run_backend_cypress.sh` and `./scripts/run_frontend_cypress.sh` separately without loading any data.
```

File: README.md [hunk 26]
```diff
@@ -16,12 +14,54 @@ Event data sets typically associate events with a certain subject (i.e. a person
...
+The frontend is accessible at http://localhost:8000 as the default "superuser" implicitly. Since the backend uses a development authentication, you can switch users by passing another users "UserId" as the access token in the query string when accessing the frontend, e.g.: http://localhost:8000/?access_token=user.user2.
+
+## Development
+
+### Testing
+Apart from separate frontend and backend tests, the project also contains end-to-end tests powered by cypress. 
+
+To run the end-to-end test locally:
+1. Make sure you installed all [requirements](#requirements)
+2. From the repo root folder run  `conquery/scripts/run_conquery_cypress.sh`
```
Comment (Kadrian) [inline README.md:58]:
```suggestion
2. From the repo root folder run `./scripts/run_conquery_cypress.sh`
```

File: README.md [hunk 27]
```diff
@@ -16,12 +14,54 @@ Event data sets typically associate events with a certain subject (i.e. a person
...
+3. Wait until the output: `Node server listening on port: 8000` appears
+4. Start cypress from the `conquery/frontend` folder in another terminal: `$ yarn cypress open`
+5. Then chose a test suite and start it.
```
Comment (Kadrian) [inline README.md:61]: This 5-steps-description is great! Tested & works like this for me :)

File: frontend/.env.e2e [hunk 28]
```diff
@@ -0,0 +1,23 @@
+# THIS IS THE ENV-FILE USED FOR CYPRESS TESTS
```
Comment (Kadrian) [inline frontend/.env.e2e:1]: 👍

**Review by Kadrian — COMMENTED** [review]
Thanks for the remaining changes! I think this PR is good to go, once you've `chmod +x`-ed the new scripts. 

I've added relative path suggestions for the README, please feel free to take or ignore those

**Review by Kadrian — DISMISSED** [review]

File: README.md (hunk 24 above)
Comment (thoniTUB) [reply README.md:36]: I thought about that, but I decided to be explicit because there are multiple `scripts` folder

**Comment by thoniTUB** [comment]
@Kadrian i had to tinker a bit. Unfortuanatelly i could not use all the scripts in the ci, just the start script
